### ISSUE TITLE
Wrong translation for sums with bounds and additional condition

### ISSUE THREAD

**Issue by wadoon** [KeYProject/key#1675]
This issue was created at [git.key-project.org](https://git.key-project.org/key/key/-/issues/1675) where the discussions are preserved.
    
----
    
## Description

When the specification contains a sum which contains bounds first and afterwards an additional condition, e.g. `(\sum int i; 0 <= i && i < 10 && 2 == 3; 1)`, the sum gets translated to a bounded sum where the additional condition is dropped.

## Reproducible
  
always

### Steps to reproduce 

1. Load the following Java class:
```java
class BsumProblem {
    
    /*@ normal_behavior
      @  ensures \result == (\sum int i; 0 <= i && i < 10 && 2 == 3; 1);
      (at)*/
    public final int m() {
        return 10;
    }
}
```

You can already see in the contract selection dialog that the additional condition `2 == 3` was dropped and a bsum occurred instead of the expected general sum. KeY verifies the contract automatically, although the sum should actually by empty due to the false additional condition.

### Additional information

* When replacing `2 == 3` by `false`, the term is translated into a general sum as expected.
* There is no warning or exception on the command line.
* Probably this can not be exploited for proving false, since the wrong term does never appear in the logic. You are only ever able to use the contract that you actually proved, it just does not match the contract you wrote down in JML. Thus I think it is not actually a soundness problem (rather some kind of a JML semantics bug).
* The problem seems to be in `JMLTermFactory` in methods `lowerBound`/`upperBound`, which in some cases extract only the bounds and drop the remaining terms.

---

* Commit: aaf908ce41012688aa4aa2caab4c0e48881aae98
    
----
    
Information: 
* created_at: 2022-01-19T15:06:14.605Z
* updated_at: 2022-01-25T15:12:25.840Z
* closed_at: None  (closed_by: )
* milestone: 
* user_notes_count: 7

**Comment by mattulbrich** [comment]
Thanks to a discussion recently, I came back to this issue.
There is a pending MR [!502](https://git.key-project.org/key/key/-/merge_requests/502) on gitlab on the topic.

I wonder if a rule could not be made such that interaction (or scripts) can be used to bound sums more generally, like

```
 sumToBsum {
    \schemaVar \variables int start, end, x;
    \schemaVar \term int t;
    \find( sum{x;}(\if(phi) \then(TRUE) \else(FALSE), t) )
    \replacewith( bsum{x;}(start, end, \if(phi) \then(t) \else(0));
    \add(==> \forall x; (phi -> start <= x <= end) )
  };
```
Such a rule could perhaps be definitorial and all other rules relating sum and bsum are consequences of this one?

We could also prove thinks like `(\sum int n; n*n <= 4; 1) == 5`.

**Comment by mattulbrich** [comment]
#3035 is a duplicate of this.

**Comment by mattulbrich** [comment]
closed by #3045
